Add topology support to ceph-csi

ShyamsundarR opened the pull request:
This PR adds Topology based volume provisioning.

The details of the design as implemented is detailed in https://github.com/ceph/ceph-csi/pull/760

ShyamsundarR commented:
Items to close beyond current code are as follows,
- Should topology be left ON by default in the provisioner sidecar?
- Should we remove topology from `volume_context` in CreateVolumeResponse
- Address 2 TODOs as follows,
  - rbd-journal.go // TODO check if need any undo operation here, or ErrVolNameConflict
  - rbd-util.go // TODO: If pool is not found we may leak the image (as DeleteVolume will return success)
- CephFS major TODO, as there is no subvolume command to fetch subvolume attributes
- Testing:
    - Topology.go needs unit test file
    - Update/add e2e as appropriate
- Manifests:
    - Helm charts update
    - Update cephfs StorageClass (later)
- Update tracevol.py
- Write documentation

humblec commented:
@ShyamsundarR are you planning to rebase it after initial review ?

Madhu-1 commented:
@ShyamsundarR will review this in the next two days

ShyamsundarR commented:
> @ShyamsundarR are you planning to rebase it after initial review ?

Yes, or sooner if I start on the TODOs before the reviews.

I am held up on some RBD mirroring testing for now, so will pick this one up a post the same.

humblec commented on cmd/cephcsi.go:
```diff
@@ -166,6 +168,13 @@ func main() {
 	}
 
+	// read and formulate conf.Domains from passed in domain labels
+	// TODO: decide if this will be from an init container and move as a separate image or instance
```
Whats the advantage we are getting it if its part of init Container ?

ShyamsundarR replied:
Using an init container keeps the ceph-csi code independent of k8s calls, and retains the CO agnostic nature of the same.

humblec commented on cmd/cephcsi.go:
```diff
@@ -166,6 +168,13 @@ func main() {
 
+	// read and formulate conf.Domains from passed in domain labels
+	// TODO: decide if this will be from an init container and move as a separate image or instance
+	conf.Domains, err = util.GetDomainsFromLabels(conf.DomainLabels, conf.NodeID)
```
Also how  are we going to deal with dynamically changed domain lables?

ShyamsundarR replied:
domain labels are set for a node, which indicate its relative location within the defined cluster domains (say rack/room/DC etc.). These are not going to change dynamically, unless the node instance is moved physically (or virtually) to another domain, at which point, I would assume, the node is restarted with the new labels, allowing us to read the new values and advertise the same.

humblec commented on pkg/util/topology.go:
```diff
@@ -0,0 +1,240 @@
+
+	// Create a k8s client to read node labels
+	client = NewK8sClient()
+	node, err := client.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
```
@ShyamsundarR this looks to be problematic to me. ooc, have we tested this PR in a setup?  reading `node` object need special permissions , Isnt it ?

ShyamsundarR replied:
There are a few ways to deal with this, for example to pass in the domain label values to the nodeplugins or create subset deployments per lowest domain with static labels. This is discussed [here](https://github.com/ceph/ceph-csi/pull/760/files#r360098011)

The manner of handling reading node labels seems the best when using daemon sets, as we need a nodeplugin running on all nodes. This does require the required roles and permissions.

Cloud providers reach into their API endpoints ([example](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/7c1a7afee2c79bdfe3f934ae8c4e839addc55720/pkg/cloud/metadata.go#L81-L106)) to get the node information, as these are used to both set the node labels (k8s) and also return CSI node domains. Further, they have access to the required credentials to make the said API calls. We lack such abilities, as we do not decide on the node labels. as these are setup by the cluster admin or the cloud provider.

In the face of the above limitations requiring additional permissions is required.

humblec commented on pkg/util/voljournal.go:
```diff
@@ -323,8 +386,15 @@ Return values:
 */
-func (cj *CSIJournal) ReserveName(ctx context.Context, monitors string, cr *Credentials, pool, reqName, parentName, encryptionKmsConfig string) (string, error) {
-	var snapSource bool
+func (cj *CSIJournal) ReserveName(ctx context.Context, monitors string, cr *Credentials,
```
in one way, the `pool` parameter now has more seggregation like `imagePool`, `journalPool`, `csiJounalPool` , imo, its better to document these concepts in source code comments for further use.

ShyamsundarR replied:
Agree, will deal with it when amending the patch for further changes.

ShyamsundarR replied:
Addressed in the latest update.

humblec commented on pkg/util/voljournal.go:
```diff
@@ -26,6 +28,8 @@ import (
 	"k8s.io/klog"
 )
 
+const uuidEncodedLength = 36
```
@ShyamsundarR anything special about the length of `36`. Just trying to understand..

ShyamsundarR replied:
This is the string representation of a UUID which is a constant length of 36 bytes ([reference](https://github.com/ceph/ceph-csi/blob/ca8dd2d8f28b6d7747330dc11b77c622ab76070f/vendor/github.com/pborman/uuid/uuid.go#L24-L28))

Unfortunately there is no constant in the uuid package to import and use, hence defined here. This needs a comment though to state it is the UUID string representation length. Will add it in the next revision to the patch set.

ShyamsundarR replied:
Addressed in the latest update.

ShyamsundarR replied:
Tried the same in go playground, it is disallowed. I could have replaced the single instance of the use with len() as above, but I **do** want to remember the length as this 36, as that was what was used in the past. So if we dynamically use len of the UUID and that changes, assumptions in code will change. Further added a note to this effect.

humblec commented:
@ShyamsundarR few review comments from my side.  One of the concern part is, how are we going to deal with dynamically changing node lables and how we can adopt the change. Have we thought about it ?

humblec commented:
`Should we remove topology from volume_context in CreateVolumeResponse`

IMO, we should not remove `tpology` from the context field in `CreateVolumeResponse` . It may be required later and no harm in keeping it.

humblec commented:
@ShyamsundarR  as we have decided to not support `allowedTopologies`, do we have to do anything here if we get that from CO ?
Additional question I have here is , how do we seperate or use the `domainLables` as specific to ceph-csi. If we read it from node objects , we are reading global domain labels and not something specific to ceph-csi and using it. Am I missing something here ?

ShyamsundarR commented:
> > Should we remove topology from volume_context in CreateVolumeResponse>
> 
> IMO, we should not remove `topology` from the context field in `CreateVolumeResponse` . It may be required later and no harm in keeping it.

Agree in general, I was worried about the volume_context bloat at present, but it is not reached a critical enough size to start reducing parameters yet from it. But, something to keep in mind.

ShyamsundarR commented:
> @ShyamsundarR as we have decided to not support `allowedTopologies`, do we have to do anything here if we get that from CO ?

The proposal [states](https://github.com/ceph/ceph-csi/pull/760/files#diff-133c3c07046f779e3502b6b5f3f1236eR74-R80) we do not support this with the current effort. There are no checks from the past or in this patch to disallow it, it will just not be parsed and used.

However, as CSI specification for CreateVolume passes only parameters for consideration, we do not get the `allowedTopologies` parameter passed in the raw to the csi-plugin. Instead the external-provisioner side car folds this into the request [like so](https://github.com/kubernetes-csi/external-provisioner/blob/efcaee79e47446e38910a3c1a024824387fcf235/pkg/controller/controller.go#L534-L547).

Essentially if we test this out, there is no further support needed from the ceph-csi implementation to support `allowedTopologies`.

> Additional question I have here is , how do we seperate or use the `domainLables` as specific to ceph-csi. If we read it from node objects , we are reading global domain labels and not something specific to ceph-csi and using it. Am I missing something here ?

`domainLabels` for ceph-csi are generated in the function [fillNodesTopology](https://github.com/ceph/ceph-csi/pull/816/files#diff-68db63fd9a7297eacb6129328d3a6ce5R57-R81). The domain labels hence are of the form `topology.<driver-name>/<domain-label>`. They carry the "value" that is the same as the the one for the node (as the node belongs to the specified domain).

The function [GetDomainsFromLabels](https://github.com/ceph/ceph-csi/pull/816/files#diff-10487ffe93db35abe03d787f0126993dR36-R40) does the required node specific prefix stripping, to provide "values" and corresponding `<domain-label>` values that is used in fillNodesTopology.

As a result, we are not mirroring the node labels, we are creating our own. Further, we need to use the nodes domain label values as we do not determine the same. Which domain labels of the node, need to be reflected by the ceph-CSI driver is controlled by the CLI args passed in with `domainlabels`

humblec commented:
@ShyamsundarR Can you revisit this PR ?

ShyamsundarR commented on cmd/cephcsi.go:
```diff
@@ -166,6 +168,13 @@ func main() {
 
+	// read and formulate conf.Domains from passed in domain labels
+	// TODO: decide if this will be from an init container and move as a separate image or instance
+	conf.Domains, err = util.GetDomainsFromLabels(conf.DomainLabels, conf.NodeID)
```
Addressed in the latest update.

ShyamsundarR commented on pkg/util/topology.go:
```diff
@@ -0,0 +1,240 @@
+	}
+
+	// Create a k8s client to read node labels
+	client = NewK8sClient()
```
Addressed in the latest update.

ShyamsundarR commented on deploy/cephfs/kubernetes/csi-cephfsplugin.yaml:
```diff
@@ -66,6 +66,12 @@ spec:
             - "--metricspath=/metrics"
             - "--enablegrpcmetrics=false"
+            # If topology based provisioning is desired, configure required
+            # kubernetes node labels representing the nodes topology domain
```
Addressed in the latest update.

ShyamsundarR commented on pkg/cephfs/volumeoptions.go:
```diff
@@ -175,6 +182,18 @@ func newVolumeOptions(ctx context.Context, requestName string, volOptions, secre
+	// TODO: we need an API to fetch subvolume attributes (size/datapool and others), based
+	// on which we can evaluate which topology this belongs to.
+	if opts.TopologyPools != nil {
+		return nil, fmt.Errorf("topology based provisioning is not supported for CephFS backed volumes")
```
Addressed in the latest update.

ShyamsundarR commented on pkg/util/topology.go:
```diff
@@ -0,0 +1,240 @@
+	// extract topology based pools configuration
+	err := json.Unmarshal([]byte(strings.Replace(topologyPoolsStr, "\n", " ", -1)), &topologyPools)
+	if err != nil {
+		klog.Infof(Log(ctx, "failed to parse JSON encoded topology constrained pools parameter (%s): %v"), topologyPoolsStr, err)
```
Addressed in the latest update.

ShyamsundarR commented on pkg/csi-common/nodeserver-default.go:
```diff
@@ -53,12 +54,47 @@ func (ns *DefaultNodeServer) NodeExpandVolume(ctx context.Context, req *csi.Node
 func (ns *DefaultNodeServer) NodeGetInfo(ctx context.Context, req *csi.NodeGetInfoRequest) (*csi.NodeGetInfoResponse, error) {
 	klog.V(5).Infof(util.Log(ctx, "Using default NodeGetInfo"))
 
+	topology, err := ns.fillNodesTopology()
```
Addressed in the latest update.

ShyamsundarR commented on pkg/csi-common/nodeserver-default.go:
```diff
@@ -53,12 +54,47 @@ func (ns *DefaultNodeServer) NodeExpandVolume(ctx context.Context, req *csi.Node
 
+	topology, err := ns.fillNodesTopology()
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "failed to update nodes topology, error : %v", err)
```
Addressed in the latest update.

Function to derive topology from domain labels is merged, and called early on to detect any errors and reused in the  NodeGetInfo function.

ShyamsundarR commented on cmd/cephcsi.go:
```diff
@@ -58,6 +58,8 @@ func init() {
 	flag.IntVar(&conf.PidLimit, "pidlimit", 0, "the PID limit to configure through cgroups")
 	flag.BoolVar(&conf.IsControllerServer, "controllerserver", false, "start cephcsi controller server")
 	flag.BoolVar(&conf.IsNodeServer, "nodeserver", false, "start cephcsi node server")
+	flag.StringVar(&conf.DomainLabels, "domainlabels", "", "list of kubernetes node labels, that determines the topology"+
```
Should we push this till we actually get another orchestrator that we work with?

ShyamsundarR commented on pkg/util/topology.go:
```diff
@@ -0,0 +1,240 @@
+/*
```
I do not have an opinion either way, but why do we need a separate folder?

ShyamsundarR commented on deploy/rbd/kubernetes/csi-nodeplugin-rbac.yaml:
```diff
@@ -3,3 +3,37 @@ apiVersion: v1
+subjects:
+  - kind: ServiceAccount
+    name: rbd-csi-nodeplugin
+    namespace: rook-ceph
```
Addressed in the latest update.

ShyamsundarR commented on deploy/cephfs/kubernetes/csi-nodeplugin-rbac.yaml:
```diff
@@ -3,3 +3,37 @@ apiVersion: v1
+subjects:
+  - kind: ServiceAccount
+    name: cephfs-csi-nodeplugin
+    namespace: rook-ceph
```
Addressed in the latest update.

ShyamsundarR commented on deploy/rbd/kubernetes/csi-rbdplugin.yaml:
```diff
@@ -66,6 +66,12 @@ spec:
             - "--metricspath=/metrics"
             - "--enablegrpcmetrics=false"
+            # If topology based provisioning is desired, configure required
+            # kubernetes node labels representing the nodes topology domain
```
Addressed in the latest update.

ShyamsundarR commented on pkg/util/topology.go:
```diff
@@ -0,0 +1,240 @@
+	}
+
+	// check if there are any accessibility requirements in the request
+	accessibilityRequirements := req.GetAccessibilityRequirements()
```
The AccessibilityRequirements is filled in by the CO, based on where the pod can be scheduled and by collecting the CSI node topology that is advertised by NodeGetInfo. IOW, user/app need not mention anything more than the topology constrained pools in the StorageClass.

ShyamsundarR commented on examples/rbd/storageclass.yaml:
```diff
@@ -48,6 +52,25 @@ parameters:
+   # are setup, and topology constrained provisioning is required.
+   # For further information read TODO<doc>
+   # topologyConstrainedPools: |
+   #   [{"poolName":"pool0",
```
This will start to get ugly, because in the topology constrained pools section we will now add one more parameter.

I forget what the limitation is to directly create the image OMap and the image on an EC pool, or is it just not supported? Tagging @dillaman as well for details.

If it is an RBD behavior, then we have no go but to add it to this section as well. But, it gets trickier than just that, we now need 2 pools per topology domain, one replicated for the OMap and image, and one EC pool for the data.

Madhu-1 replied:
Need clarification on this one.

ShyamsundarR commented:
> we may need to add below
> 
>     * A Readme doc topology-based provisioning in CSI (examples)
> 
>     * An E2E just to cover basic scenario (once we Fix #857 we can extend it to multi-node tests)

Tracked in the TODO list present in this [comment](https://github.com/ceph/ceph-csi/pull/816#issuecomment-586775091)

Madhu-1 commented on pkg/cephfs/fsjournal.go:
```diff
@@ -46,52 +46,44 @@ request name lock, and hence any stale omaps are leftovers from incomplete trans
+	// check if topology constraints match what is found
+	// TODO: we need an API to fetch subvolume attributes (size/datapool and others), based
+	// on which we can evaluate which topology this belongs to.
+	// TODO: CephFS topology support is postponed till we get the same
```
is this true? because I see template  changes for cephfs also

ShyamsundarR replied:
Everything else is in place, **except** the ability to inspect an subvolumes data pool. Hence, leaving it in place and will cover support/feature situation for CephFS in documentation.

Madhu-1 commented on pkg/rbd/driver.go:
```diff
@@ -133,8 +133,16 @@ func (r *Driver) Run(conf *util.Config, cachePersister util.CachePersister) {
+		}
+	}
+
+	if conf.IsNodeServer {
```
we can remove this change and move below code to line 142

ShyamsundarR replied:
Addressed in the latest update.

Madhu-1 commented on pkg/rbd/driver.go:
```diff
@@ -144,7 +152,7 @@ func (r *Driver) Run(conf *util.Config, cachePersister util.CachePersister) {
 	}
 	if !conf.IsControllerServer && !conf.IsNodeServer {
-		r.ns, err = NewNodeServer(r.cd, conf.Vtype)
+		r.ns, err = NewNodeServer(r.cd, conf.Vtype, topology)
```
the topology will be empty here isn't it?

ShyamsundarR replied:
Addressed in the latest update. Good catch!

Madhu-1 commented on scripts/rook-rbd-pool.sh:
```diff
@@ -0,0 +1,54 @@
+ROOK_URL="https://raw.githubusercontent.com/rook/rook/${ROOK_VERSION}/cluster/examples/kubernetes/ceph"
+ROOK_RBD_POOL=${ROOK_RBD_POOL:-"newrbdpool"}
+
+function create_pool() {
```
move this code changes to rook.sh file

ShyamsundarR replied:
Addressed in the latest update.

Madhu-1 commented on scripts/rook-rbd-pool.sh:
```diff
@@ -0,0 +1,54 @@
+function delete_pool() {
+	curl -o newpool.yaml "${ROOK_URL}/pool-test.yaml"
+    sed -i "s/replicapool/$ROOK_RBD_POOL/g" newpool.yaml
+	kubectl delete -f "./newpool.yaml"
```
we also need to delete newpool.yaml file better to save it in tmp folder

ShyamsundarR replied:
`tmp` folder save did not work for me, due to (possible) minikube settings, hence I decided to just dump it in place. If this is fine, we can look at this later.

ShyamsundarR commented on charts/ceph-csi-cephfs/templates/provisioner-deployment.yaml:
```diff
@@ -37,6 +37,9 @@ spec:
             - "--leader-election-type=leases"
             - "--retry-interval-start=500ms"
+{{- if .Values.topology.enabled }}
+            - "--feature-gates=Topology=true"
```
TODO: Documentation should cover versions and support information, will keep that in mind.

The support status is as [here](https://kubernetes-csi.github.io/docs/topology.html#status), IOW this needs kubelet to be v1.14 or greater, and the external-provisioner side car to be 1.1 or greater. Further, there is the kubernetes feature gate for versions lower than GA as covered [here](https://kubernetes-csi.github.io/docs/topology.html#beta)

ShyamsundarR commented on charts/ceph-csi-rbd/templates/nodeplugin-clusterrole.yaml:
```diff
@@ -0,0 +1,19 @@
+{{- if .Values.rbac.create -}}
+{{- if .Values.topology.enabled }}
```
True, when we come to that point we need to change the if conditions appropriately, at present as you/we cleaned up the RBAC, the entire file is either needed or not. So, I used an overarching if block here.

ShyamsundarR commented on charts/ceph-csi-cephfs/templates/nodeplugin-clusterrole.yaml:
```diff
@@ -0,0 +1,19 @@
+        rbac.cephfs.csi.ceph.com/aggregate-to-{{ include "ceph-csi-cephfs.nodeplugin.fullname" . }}: "true"
+rules: []
+{{- end }}
+{{- end -}}
```
Addressed in the latest update.

ShyamsundarR commented on e2e/utils.go:
```diff
@@ -1119,3 +1145,120 @@ func checkMountOptions(pvcPath, appPath string, f *framework.Framework, mountFla
+	if err != nil {
+		Fail(err.Error())
+	}
+	for _, expression := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms[0].MatchExpressions {
```
Addressed in the latest update.

Madhu-1 commented on scripts/rook.sh:
```diff
@@ -115,11 +150,19 @@ deploy)
+  deploy             Deploy a rook
+  teardown           Teardown a rook
+  create-block-pool  Create a rook block pool
+  delete-block-pool  Delete a rook block pool
```
are we using delete-block-pool any where?

ShyamsundarR replied:
no, provided for completeness.

Madhu-1 commented on scripts/rook.sh:
```diff
@@ -37,6 +38,40 @@ function teardown_rook() {
+	curl -o newpool.yaml "${ROOK_URL}/pool-test.yaml"
+    sed -i "s/replicapool/$ROOK_BLOCK_POOL_NAME/g" newpool.yaml
+	kubectl delete -f "./newpool.yaml"
+	rm -f "./newpool.yaml"
```
do we need to check in toolbox that pool is deleted?

ShyamsundarR replied:
This is provided for completeness, so not adding any checks as such (like create pool)

Madhu-1 commented on scripts/rook.sh:
```diff
@@ -37,6 +38,40 @@ function teardown_rook() {
+	kubectl create -f "./newpool.yaml"
+	rm -f "./newpool.yaml"
+
+    for ((retry = 0; retry <= ROOK_DEPLOY_TIMEOUT; retry = retry + 5)); do
+		echo "Checking RBD ($ROOK_BLOCK_POOL_NAME) stats... ${retry}s" && sleep 5
+
+		TOOLBOX_POD=$(kubectl -n rook-ceph get pods -l app=rook-ceph-tools -o jsonpath='{.items[0].metadata.name}')
+		TOOLBOX_POD_STATUS=$(kubectl -n rook-ceph get pod "$TOOLBOX_POD" -ojsonpath='{.status.phase}')
+		[[ "$TOOLBOX_POD_STATUS" != "Running" ]] && \
+			{ echo "Toolbox POD ($TOOLBOX_POD) status: [$TOOLBOX_POD_STATUS]"; continue; }
+
+		if kubectl exec -n rook-ceph "$TOOLBOX_POD" -it -- rbd pool stats "$ROOK_BLOCK_POOL_NAME" &>/dev/null; then
+			echo "RBD ($ROOK_BLOCK_POOL_NAME) is successfully created..."
+			break
+		fi
+	done
+
+	if [ "$retry" -gt "$ROOK_DEPLOY_TIMEOUT" ]; then
+		echo "[Timeout] Failed to get RBD pool $ROOK_BLOCK_POOL_NAME stats"
+		exit 1
+	fi
+	echo ""
+}
```
can we avoid the code duplication here currently we have function to check the pool stats. i think we can have one function which accepts input name as pool name and check in toolbox pod https://github.com/ceph/ceph-csi/blob/master/scripts/rook.sh#L88-L109

ShyamsundarR replied:
Addressed in the latest update.

Madhu-1 commented:
@humblec @nixpanic PTAL

phlogistonjohn commented on cmd/cephcsi.go:
```diff
@@ -58,6 +58,8 @@ func init() {
 	flag.BoolVar(&conf.IsControllerServer, "controllerserver", false, "start cephcsi controller server")
 	flag.BoolVar(&conf.IsNodeServer, "nodeserver", false, "start cephcsi node server")
+	flag.StringVar(&conf.DomainLabels, "domainlabels", "", "list of kubernetes node labels, that determines the topology"+
+		" domain the node belongs to, separated by ';'")
```
+1 to that and pflag already has support for doing that kind of thing baked in:
https://godoc.org/github.com/spf13/pflag#StringSliceVar

ShyamsundarR replied:
Addressed in the latest update.

I did not use `StringSliceVar` from spf13/pflag as suggested by @phlogistonjohn  as the current flag package in use, seems to be the one that is default with go (although we do have the mentioned package in the vendor directory). Reason being, I did not want to add a new dependency with this commit.

phlogistonjohn replied:
Oh, I was unaware. Sorry for the noise. I just assumed everyone was using pflag now. :-)

ShyamsundarR replied:
@phlogistonjohn Suggestion was useful, I almost did 3 lines of change, before hunting for the API in the **current** flag package, so no worries. We should move to pflag is that is more appropriate in the future.

ShyamsundarR commented on charts/ceph-csi-cephfs/values.yaml:
```diff
@@ -165,6 +165,17 @@ provisioner:
+  # Each node label is ; separated in the value below
+  # NOTE: the value here serves as an example and needs to be
+  # updated with node labels that define domains of interest
+  domainLabels: failure-domain/region;failure-domain/zone
```
Addressed in the latest update.

nixpanic commented on scripts/rook.sh:
```diff
@@ -37,6 +38,22 @@ function teardown_rook() {
+
+function delete_block_pool() {
+	curl -o newpool.yaml "${ROOK_URL}/pool-test.yaml"
+    sed -i "s/replicapool/$ROOK_BLOCK_POOL_NAME/g" newpool.yaml
```
indention?

ShyamsundarR replied:
Addressed in the latest update.

nixpanic commented on scripts/rook.sh:
```diff
@@ -37,6 +38,22 @@ function teardown_rook() {
 
+function create_block_pool() {
+    curl -o newpool.yaml "${ROOK_URL}/pool-test.yaml"
+    sed -i "s/replicapool/$ROOK_BLOCK_POOL_NAME/g" newpool.yaml
```
indention?

ShyamsundarR replied:
Addressed in the latest update.

ShyamsundarR commented on pkg/rbd/rbd_util.go:
```diff
@@ -394,17 +413,29 @@ func genVolFromVolID(ctx context.Context, rbdVol *rbdVolume, volumeID string, cr
 	if err != nil {
 		return err
 	}
+	rbdVol.JournalPool = rbdVol.Pool
```
Added a comment on what `DataPool` is, and where it is used (only during creates) in the `rbdVolume` struct.

When generating the volume from the VolID, we do not need to populate the `DataPool` as was the case prior to this change.

ShyamsundarR commented on pkg/util/voljournal.go:
```diff
@@ -198,71 +219,114 @@ Return values:
+		if err != nil {
+			return nil, err
+		}
+		savedImagePoolID = int64(binary.BigEndian.Uint64(buf64))
```
We store this as BigEndian at line 516 in the same file

nixpanic replied:
Yes, I noticed that later too. Is there a reason for BigEndian, and not just a string?

ShyamsundarR replied:
This is an pool ID that is an int64 type, so prefer to retain this encoded. The same encoding is used in the `volumeID` as well, when I wrote that up.

ShyamsundarR commented on pkg/util/voljournal.go:
```diff
@@ -366,54 +455,78 @@ func (cj *CSIJournal) ReserveName(ctx context.Context, monitors string, cr *Cred
-	// Update UUID directory to store CSI request name and image name
-	err = SetOMapKeyValue(ctx, monitors, cr, pool, cj.namespace, cj.cephUUIDDirectoryPrefix+volUUID,
+	// Update UUID directory to store CSI request name
+	err = SetOMapKeyValue(ctx, monitors, cr, imagePool, cj.namespace, cj.cephUUIDDirectoryPrefix+volUUID,
```
The image journal, or OMap directory is stored with the image, which goes to the `imagePool`.

This was due to the reasons [here](https://github.com/ceph/ceph-csi/pull/760/files#r368085409). Added a comment in the code regarding the same.

ShyamsundarR commented on scripts/minikube.sh:
```diff
@@ -118,6 +121,16 @@ deploy-rook)
     DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
     "$DIR"/rook.sh deploy
     ;;
+create-block-pool)
```
This is a glue between `rook.sh` and `minikube.sh`, just so that the travis scripts only invoke `minikube.sh`. I would assume their presence in rook.sh helps running them on non-minikube environments.

ShyamsundarR commented on pkg/util/topology.go:
```diff
@@ -0,0 +1,257 @@
+	topology := make(map[string]string)
+	for domain, value := range domainMap {
+		topology[topologyPrefix+"/"+domain] = value
+		// TODO: when implementing domain takeover/giveback, enable a domain value that can remain pinned to the node
```
Update the TODO list in the tracking [comment](https://github.com/ceph/ceph-csi/pull/816#issuecomment-586775091). I will open issues for these once the PR merges.

ShyamsundarR commented on pkg/util/topology.go:
```diff
@@ -0,0 +1,257 @@
+
+// FindPoolAndTopology loops through passed in "topologyPools" and also related
+// accessibility requirements, to determine which pool matches the requirement.
+// The return variables are, poolname, topology map of matched requirement
```
Addressed in the latest update.

ShyamsundarR commented on pkg/cephfs/volumeoptions.go:
```diff
@@ -176,6 +183,18 @@ func newVolumeOptions(ctx context.Context, requestName string, volOptions, secre
+	}
+
+	// TODO: we need an API to fetch subvolume attributes (size/datapool and others), based
+	// on which we can evaluate which topology this belongs to.
```
Addressed in the latest update.

ShyamsundarR commented on pkg/rbd/rbd_util.go:
```diff
@@ -70,24 +71,28 @@ type rbdVolume struct {
+	Monitors            string `json:"monitors"`
+	JournalPool         string
+	Pool                string `json:"pool"`
+	DataPool            string
```
Addressed in the latest update.

ShyamsundarR commented on pkg/util/topology.go:
```diff
@@ -0,0 +1,257 @@
+
+// GetTopologyFromDomainLabels returns the CSI topology map, determined from
+// the domain labels and their values from the CO system
+// Expects domainLabels in arg to be in the format "[prefix/]<name>,[prefix/]<name>;...",
```
Addressed in the latest update.

ShyamsundarR commented on pkg/util/topology.go:
```diff
@@ -0,0 +1,257 @@
+
+	// Ensure all labels are found
+	if found != labelCount {
+		missingLabels := ""
```
Addressed in the latest update.

ShyamsundarR commented on pkg/util/voljournal.go:
```diff
@@ -423,70 +536,89 @@ func (cj *CSIJournal) ReserveName(ctx context.Context, monitors string, cr *Cred
+	SourceName    string // Contains the parent image name for the passed in UUID, if it is a snapshot
+	ImageName     string // Contains the image or subvolume name for the passed in UUID
+	KmsID         string // Contains encryption KMS, if it is an encrypted image
+	JournalPoolID int64
```
Addressed in the latest update.

ShyamsundarR commented on scripts/install-helm.sh:
```diff
@@ -130,6 +136,13 @@ install_cephcsi_helm_charts() {
     fi
+
+    # label the nodes uniformly for domain information
+    for node in $(kubectl get node --no-headers | cut -f 1 -d ' '); do
```
Addressed in the latest update.

nixpanic commented on e2e/utils.go:
```diff
@@ -319,6 +319,12 @@ func createRBDStorageClass(c kubernetes.Interface, f *framework.Framework, param
 	}
 	sc.Namespace = cephCSINamespace
+	for k, v := range scOptions {
+		if k == "volumeBindingMode" && v == "WaitForFirstConsumer" {
```
Are these strings not available as constants?

ShyamsundarR replied:
I cannot find a reusable constant for `volumeBindingMode`.

For `WaitForFirstConsumer` there is a typed string constant (which is used just below this as `scv1.VolumeBindingWaitForFirstConsumer`), but this makes them non type compliant causing linter failures.

Also, this means we need to map for `scOptions` that this function takes, be a map with different keys and values?

If you have an alternative I can code that up, else I am out of options on this.

humblec replied:
declared `type VolumeBindingMode string` constants can be used here @ShyamsundarR instead of hard-coding these.

ShyamsundarR replied:
As stated, I am using these constants just below, but due to type difference not able to cleanly use them here. I would assume this does not stop the PR from being merged, though. If it does, please let me know.

nixpanic commented:
Looks pretty good and complete to me. Just a few more minor things and I'd give it a :+1:

Madhu-1 commented:
@nixpanic can you please approve this if all your comments are resolved?

Madhu-1 commented on e2e/utils.go:
```diff
@@ -319,6 +319,12 @@ func createRBDStorageClass(c kubernetes.Interface, f *framework.Framework, param
 		sc.Parameters[k] = v
 	}
 	sc.Namespace = cephCSINamespace
+	for k, v := range scOptions {
+		if k == "volumeBindingMode" && v == "WaitForFirstConsumer" {
+			value := scv1.VolumeBindingWaitForFirstConsumer
+			sc.VolumeBindingMode = &value
+		}
+	}
```
```suggestion
if scOptions["volumeBindingMode"]=="WaitForFirstConsumer"{
value := scv1.VolumeBindingWaitForFirstConsumer
sc.VolumeBindingMode = &value
}
```

ShyamsundarR replied:
Updated in the latest rebase.

humblec commented on deploy/cephfs/kubernetes/csi-cephfsplugin-provisioner.yaml:
```diff
@@ -40,6 +40,7 @@ spec:
             - "--enable-leader-election=true"
             - "--leader-election-type=leases"
             - "--retry-interval-start=500ms"
+            - "--feature-gates=Topology=true"
```
@ShyamsundarR do we need to explicitly have a parameter here wih `true` value ?

ShyamsundarR replied:
yes, see [this](https://kubernetes-csi.github.io/docs/topology.html#sidecar-deployment) document for the same.

humblec commented on docs/deploy-cephfs.md:
```diff
@@ -44,22 +44,23 @@ that should be resolved in v14.2.3.
 | `--metadatastorage`       | _empty_                     | Points to where older (1.0.0 or older plugin versions) metadata about provisioned volumes are kept, as file or in as k8s configmap (`node` or `k8s_configmap` respectively)                                                                                                          |
 | `--pidlimit`              | _0_                         | Configure the PID limit in cgroups. The container runtime can restrict the number of processes/tasks which can cause problems while provisioning (or deleting) a large number of volumes. A value of `-1` configures the limit to the maximum, `0` does not configure limits at all. |
-| `--metricsport`           | `8080`                      | TCP port for liveness metrics requests                                                                                                                                                                                                                                                  |
+| `--metricsport`           | `8080`                      | TCP port for liveness metrics requests                                                                                                                                                                                                                                               |
```
@ShyamsundarR whats the change here for `metricsport` and `enable...` ?  not able to figure out.

ShyamsundarR replied:
Merge issue, these need to be removed. Working on it now.

ShyamsundarR replied:
Ah! ok, this was not a merge issue after all (which is a relief).

I have seen that whenever we modify YAML tables we are aligning these to the same width. The same was done here post addition of the new option. Although the new option did not change the table width, other added options that did not adjust for the table width (in their commits) were now auto-corrected. IOW, the trailing `|` is now auto-corrected to the right location, which is the change here.

Also, when reviewing this, noticed that the block `[rbd | cephfs]` in option --type was incorrectly aligned due to the table width correction, which is now reverted.

humblec commented on pkg/util/topology_test.go:
```diff
@@ -0,0 +1,379 @@
+}
+
+/*
+// TODO: To test GetTopologyFromDomainLabels we need it to accept a k8s client interface, to mock k8sGetNdeLabels output
```
@ShyamsundarR is this TODO planned for follow up PR ?

ShyamsundarR replied:
It is left in here if decided to take this up as an InitContainer.

ShyamsundarR replied:
IOW, no followup PR planned for this.

humblec commented:
Overal looks good to me. some minor comments from end

humblec commented:
@nixpanic any final comments from your side ?

humblec commented:
LGTM.. Thanks ! @ShyamsundarR can you please do  rebase ?

nixpanic commented on pkg/util/topology.go:
```diff
@@ -0,0 +1,255 @@
+
+package util
+
+import (
```
Add `"context"`

nixpanic commented on pkg/util/topology.go:
```diff
@@ -0,0 +1,255 @@
+
+func k8sGetNodeLabels(nodeName string) (map[string]string, error) {
+	client := NewK8sClient()
+	node, err := client.CoreV1().Nodes().Get(nodeName, metav1.GetOptions{})
```
With the upgrade to k8s 1.18.0 this needs an extra parameter, probably `context.TODO()`

ShyamsundarR commented:
Rebased to pull in the latest k8s 1.18 changes. Had to do the following,
- Added function getBoundPV as that is no longer available from the k8s framework package see [this](https://github.com/ceph/ceph-csi/pull/816/files#diff-0781e6445e81ef5ec3953f597be8dc1aR1311-R1316)
- Updated k8s calls to add the `context.TODO()` as appropriate (in e2e/utils.go and in topology.go)

Ran only build and static checks on my development machine. Will await e2e results for a final check.

humblec commented:
> Rebased to pull in the latest k8s 1.18 changes. Had to do the following,
> 
>     * Added function getBoundPV as that is no longer available from the k8s framework package see [this](https://github.com/ceph/ceph-csi/pull/816/files#diff-0781e6445e81ef5ec3953f597be8dc1aR1311-R1316)
> 
>     * Updated k8s calls to add the `context.TODO()` as appropriate (in e2e/utils.go and in topology.go)
> 
> 
> Ran only build and static checks on my development machine. Will await e2e results for a final check.

Thanks @ShyamsundarR . Just waiting for the e2e to pass !

humblec commented:
All the tests are green. LGTM .. Approved.!
